README.md
# toprogress2

[![Version][npm-version-shield]][npm]
![License][mit-shield]
[![Build Status][travis-ci-shield]][travis-ci]
[![Coverage States][codecov-shield]][codecov]
[![Downloads][npm-stats-shield]][npm-stats]

[npm]: https://www.npmjs.com/package/toprogress2
[npm-version-shield]: https://img.shields.io/npm/v/toprogress2.svg

[mit-shield]: https://img.shields.io/npm/l/toprogress2.svg

[travis-ci]: https://travis-ci.org/caseyWebb/toprogress2/
[travis-ci-shield]: https://img.shields.io/travis/caseyWebb/toprogress2/master.svg

[codecov]: https://codecov.io/gh/caseyWebb/toprogress2
[codecov-shield]: https://img.shields.io/codecov/c/github/caseyWebb/toprogress2.svg

[npm-stats]: http://npm-stat.com/charts.html?package=toprogress2&author=&from=&to=
[npm-stats-shield]: https://img.shields.io/npm/dt/toprogress2.svg?maxAge=2592000

Forked from [djyde/ToProgress](https://github.com/djyde/ToProgress). See [CHANGELOG.md](./CHANGELOG.md).

![](https://caseyWebb.github.io/toprogress2/screenshot.gif)

# Quick Start
```bash
$ yarn add toprogress2
```

```JS
import { ToProgress } from 'toprogress2'

const topbar = new ToProgress({
  id: 'toprogress',
  color: '#F44336',
  height: '2px',
  duration: 0.2,
  position: 'top',
  selector: 'body'
})

topbar.start()
topbar.stop()
topbar.finish()
```

# API

Documentation is generated via Typedoc and is available [here](https://caseyWebb.github.io/toprogress2/classes/toprogress.html)

# Demo

View the demo [here](https://caseyWebb.github.io/toprogress2/demo/)

# Browser Support

* IE >= 10
* Chrome
* Firefox
* Safari

# License
MIT License


CHANGELOG.md
# 2.0.0

### Added
- `.start()` method
- `.stop()` method
- `.destroy()` method

### Changed
- [BREAKING] Replaced all callback functions with promises


# 1.0.0

### Added
- TypeScript definitions

### Changed
- Moved `selector` argument to options object

### Internal
- Rewrite in TypeScript
- Add unit tests

test.ts
import { ToProgress } from './index'

function emitTransitionEndEvent(style: string) {
  const e: any = new Event('transitionend')
  const el = document.getElementById('toprogress')
  e.propertyName = style
  el.dispatchEvent(e)
}

beforeEach(() => {
  document.body.innerHTML = ''
})

describe('options', () => {
  describe('id', () => {
    test('defaults to toprogress', () => {
      const bar = new ToProgress()
      expect(document.getElementById('toprogress')).toBeTruthy()
    })
    test('sets progress bar element id', () => {
      const bar = new ToProgress({ id: 'progressbar' })
      expect(document.getElementById('progressbar')).toBeTruthy()
    })
  })

  describe('selector', () => {
    test('defaults to body', () => {
      const bar = new ToProgress()
      expect(document.body.firstChild).toEqual(document.getElementById('toprogress'))
    })
    test('throws if element does not exist', () => {
      expect(() => {
        const bar = new ToProgress({ selector: 'dne' })
      }).toThrow()
    })
    test('inserts progress bar as first child of empty element', () => {
      const container = document.createElement('div')
      container.id = 'container'
      document.body.appendChild(container)

      const bar = new ToProgress({ selector: '#container' })
      expect(container.firstChild).toEqual(document.getElementById('toprogress'))
    })
    test('inserts progress bar as first child of non-empty element', () => {
      const container = document.createElement('div')
      container.id = 'container'
      container.appendChild(document.createElement('div'))
      document.body.appendChild(container)

      const bar = new ToProgress({ selector: '#container' })

      expect(container.firstChild).toEqual(document.getElementById('toprogress'))
    })
    test('sets position using absolute positioning', () => {
      const container = document.createElement('div')
      container.id = 'container'
      container.appendChild(document.createElement('div'))
      document.body.appendChild(container)
      const bar = new ToProgress({ selector: '#container' })
      const el = document.getElementById('toprogress')
      expect(el.style.position).toBe('absolute')
      expect(el.style.top).toBe('0px')
      expect(el.style.bottom).toBeFalsy()
    })
  })

  describe('color', () => {
    test('defaults to #F44336', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      expect(el.style.backgroundColor).toBe('rgb(244, 67, 54)')
    })
    test('sets element backgroundColor', () => {
      const bar = new ToProgress({ color: '#FFF' })
      const el = document.getElementById('toprogress')
      expect(el.style.backgroundColor).toBe('rgb(255, 255, 255)')
    })
  })

  describe('height', () => {
    test('defaults to 2px', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      expect(el.style.height).toBe('2px')
    })
    test('sets element height', () => {
      const bar = new ToProgress({ height: '5px' })
      const el = document.getElementById('toprogress')
      expect(el.style.height).toBe('5px')
    })
  })

  describe('duration', () => {
    test('defaults to 0.2s', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      expect(el.style.transition).toContain('width 0.2s')
    })
    test('sets duration', () => {
      const bar = new ToProgress({ duration: 5 })
      const el = document.getElementById('toprogress')
      expect(el.style.transition).toContain('width 5s')
    })
    test('opacity duration is 3 x duration', () => {
      const bar = new ToProgress({ duration: .1 })
      const el = document.getElementById('toprogress')
      expect(el.style.transition).toContain(`opacity ${.1 * 3}s`)
    })
    test('max opacity duration is 1 second', () => {
      const bar = new ToProgress({ duration: 5 })
      const el = document.getElementById('toprogress')
      expect(el.style.transition).toContain('opacity 1s')
    })
  })

  describe('position', () => {
    test('defaults to top', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      expect(el.style.position).toBe('fixed')
      expect(el.style.top).toBe('0px')
      expect(el.style.bottom).toBeFalsy()
    })
    test('sets position using fixed positioning', () => {
      const bar = new ToProgress({ position: 'bottom' })
      const el = document.getElementById('toprogress')
      expect(el.style.position).toBe('fixed')
      expect(el.style.top).toBeFalsy()
      expect(el.style.bottom).toBe('0px')
    })
  })
})

describe('initialization', () => {
  test('progress starts at 0', () => {
    const bar = new ToProgress()
    const el = document.getElementById('toprogress')
    expect(el.style.width).toBe('0%')
  })
  test('opacity is initialized as 1', () => {
    const bar = new ToProgress()
    const el = document.getElementById('toprogress')
    expect(el.style.opacity).toBe('1')
  })
})

describe('api', () => {
  describe('.start()', () => {
    test('uses default 30s duration and cubic-beizer', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.start()
      expect(el.style.transition).toContain('width 30s cubic-bezier')
    })

    test('can set duration', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.start(10)
      expect(el.style.transition).toContain('width 10s cubic-bezier')
    })

    test('can set easing', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.start(10, 'linear')
      expect(el.style.transition).toContain('width 10s linear')
    })

    test('sets the width to 99', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.start()
      expect(el.style.width).toBe('99%')
    })

    test('calling repeatedly re-triggers css transition', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.start()
      expect(el.style.width).toBe('99%')
      bar.start()
      expect(el.style.width).toBe('98%')
      bar.start()
      expect(el.style.width).toBe('99%')
    })
  })

  describe('.stop()', () => {
    test('sets the transition duration to a big number', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.stop()
      expect(el.style.transition).toContain('100000')
    })

    test('re-triggers the animation', () => {
      const bar = new ToProgress()
      bar.start()
      bar.stop()
      expect(bar.getProgress()).not.toBe(99)
    })
  })

  describe('.setProgress()', () => {
    test('sets the width', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.setProgress(50)
      expect(el.style.width).toBe('50%')
    })

    test('returns a promise that resolves after transitionEnd', (done) => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.setProgress(10).then(() => {
        expect(el.style.width).toBe('10%')
        done()
      })
      emitTransitionEndEvent('width')
    })

    test('sets width to 100% when n > 100', () => {
      const bar = new ToProgress()
      bar.setProgress(110)
      expect(bar.getProgress()).toBe(100)
    })

    test('sets width to 0% when n < 0', () => {
      const bar = new ToProgress()
      bar.setProgress(-1)
      expect(bar.getProgress()).toBe(0)
    })
  })

  describe('.increase() / .decrease()', () => {
    test('increment/decrement the progress', () => {
      const bar = new ToProgress()
      bar.increase(10)
      expect(bar.getProgress()).toBe(10)
      bar.decrease(5)
      expect(bar.getProgress()).toBe(5)
    })
  })

  describe('.hide() / .show()', () => {
    test('toggle the opacity', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.hide()
      expect(el.style.opacity).toBe('0')
      bar.show()
      expect(el.style.opacity).toBe('1')
    })
  })

  describe('.getProgress()', () => {
    test('returns the current progress', () => {
      const bar = new ToProgress()
      bar.setProgress(50)
      expect(bar.getProgress()).toBe(50)
    })
  })

  describe('.reset()', () => {
    test('sets the progress to 0', () => {
      const bar = new ToProgress()
      bar.setProgress(50)
      bar.reset()
      expect(bar.getProgress()).toBe(0)
    })

    test('sets the transition duration and easing to default', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.reset()
      expect(el.style.transition).toContain('width 0.2s ease-out')
    })

    test('calls .show()', (done) => {
      const bar = new ToProgress()
      bar.show = jest.fn()
      bar.reset().then(() => {
        expect(bar.show).toBeCalled()
        done()
      })
      emitTransitionEndEvent('width')
    })

    test('returns a promise that resolves after transitionEnd', (done) => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.reset().then(done)
      emitTransitionEndEvent('width')
    })
  })

  describe('.finish()', () => {
    test('sets the progress to 100%', () => {
      const bar = new ToProgress()
      bar.finish()
      expect(bar.getProgress()).toBe(100)
    })

    test('sets the transition duration and easing to default', () => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.finish()
      expect(el.style.transition).toContain('width 0.2s ease-out')
    })

    test('calls .hide()', () => {
      const bar = new ToProgress()
      bar.hide = jest.fn()
      bar.finish()
      expect(bar.hide).toBeCalled()
    })

    test('calls .reset() after opacity transition completes', (done) => {
      const bar = new ToProgress()
      const el = document.getElementById('toprogress')
      bar.reset = jest.fn()
      bar.finish()
      emitTransitionEndEvent('opacity')
      process.nextTick(() => {
        expect(bar.reset).toBeCalled()
        done()
      })
    })
  })

  describe('.destroy()', () => {
    test('removes the element from the DOM', () => {
      const bar = new ToProgress()
      bar.destroy()
      expect(document.getElementById('toprogress')).toBeFalsy()
    })
  })
})


index.ts
function whichTransitionEvent() {
  const el = document.createElement('fakeelement')
  const transitions: { [k: string]: string } = {
    OTransition     : 'oTransitionEnd',
    WebkitTransition: 'webkitTransitionEnd'
  }
  for (const t in transitions) {
    /* istanbul ignore next */
    if ((el.style as any)[t] !== undefined) {
      /* istanbul ignore next */
      return transitions[t]
    }
  }
  return 'transitionend'
}

const transitionEvent = whichTransitionEvent()

// tslint:disable-next-line interface-name
export interface ToProgressOptions {
  /**
   * id attribute to assign progress bar element
   * @default `'toprogress'`
   */
  id?: string
  /**
   * Progress bar color
   * @default `'#F44336'`
   */
  color?: string
  /**
   * Progress bar height
   * @default `'2px'`
   */
  height?: string
  /**
   * CSS transition duration for `.finish()`, `.setProgress()`, `.increase()`, `.decrease()`, and `.reset()`
   * @default `'0.2s'`
   */
  duration?: number
  /**
   * Progress bar position. `top` or `bottom`
   * @default `'top'`
   */
  position?: string
  /**
   * Selector of progress bar container
   * @default `'body'`
   */
  selector?: string
}

export class ToProgress {
  private options = {
    id: 'toprogress',
    color: '#F44336',
    height: '2px',
    duration: 0.2,
    position: 'top',
    selector: 'body'
  }
  private element = document.createElement('div')
  private progress = 0

  constructor(opts?: ToProgressOptions) {
    Object.assign(this.options, opts || {})
    this.element.id = this.options.id
    this.setCSS()
    this.setTransition(this.options.duration)
    this.createProgressBar()
  }

  /**
   * Note, because .start() uses a single CSS transition, this will not
   * accurately reflect the position on the screen
   * @returns Current progress
   */
  public getProgress() {
    return this.progress
  }

  /**
   * @param progress 0-100
   * @returns Promise which resolves after transition animation completes
   */
  public setProgress(progress: number) {
    this.progress = Math.min(100, Math.max(0, progress))
    this.element.style.width = this.progress + '%'
    return this.transitionEnd('width')
  }

  /**
   * @param duration Max amount of time to animate. Larger values = slower moving bar.
   * @param easing CSS transition easing
   * @returns Promise which resolves after timeout or `.finish()`
   */
  public start(duration: number = 30, easing = 'cubic-bezier(0.05, 0.45, 0.05, 0.95)') {
    this.show()
    this.setTransition(duration, easing)
    /**
     * Need to ensure that if already started the width is changed so
     * that the new transition takes effect.
     */
    return this.setProgress(this.progress === 99 ? 98 : 99)
  }

  public stop() {
    this.setTransition(10e10)
    this.setProgress(this.progress - 1).catch(() => { /* noop */})
  }

  /**
   * @param i Amount to increment progress
   * @returns Promise which resolves after animation completes
   */
  public increase(i: number) {
    return this.setProgress(this.progress + i)
  }

  /**
   * @param i Amount to decrement progress
   * @returns Promise which resolves after animation completes
   */
  public decrease(i: number) {
    return this.setProgress(this.progress - i)
  }

  public hide() {
    this.element.style.opacity = '0'
  }

  public show() {
    this.element.style.opacity = '1'
  }

  /**
   * Transition to 100%, hide, and reset
   *
   * @returns Promise which resolves after progress bar has been reset
   */
  public finish() {
    this.setTransition(this.options.duration)
    this.setProgress(100).catch(() => { /* noop */ })
    this.hide()
    return this.transitionEnd('opacity').then(() => this.reset())
  }

  /**
   * Set progress to 0 and show after transition completes
   *
   * @returns Promise which resolves after progress bar has been reset
   */
  public reset() {
    this.setTransition(this.options.duration)
    return this.setProgress(0).then(() => this.show())
  }

  public destroy() {
    this.element.remove()
  }

  private setCSS() {
    const styles: { [k: string]: string } = {
      'position': this.options.selector === 'body' ? 'fixed' : 'absolute',
      'top': this.options.position === 'top' ? '0' : 'auto' ,
      'bottom': this.options.position === 'bottom' ? '0' : 'auto' ,
      'left': '0',
      'right': '0',
      'background-color': this.options.color,
      'height': this.options.height,
      'width': '0%',
      'opacity': '1',
      'box-shadow': `0px 1px 2px 0px ${this.options.color}`
    }
    Object.keys(styles).forEach((style: string) => (this.element.style as any)[style] = styles[style])
  }

  private setTransition(duration: number, easing = 'ease-out') {
    const opacityDuration = Math.min(duration * 3, 1)
    const rule = `width ${duration}s ${easing}, opacity ${opacityDuration}s`
    const styles: { [k: string]: string } = {
      'transition': rule,
      '-moz-transition': rule,
      '-webkit-transition': rule
    }
    Object.keys(styles).forEach((style: string) => (this.element.style as any)[style] = styles[style])
  }

  private createProgressBar() {
    const el = document.querySelector(this.options.selector)
    if (!el) {
      throw new Error(`[toprogress2] Element not found with selector ${this.options.selector}`)
    }
    if (el.hasChildNodes()) {
      el.insertBefore(this.element, el.firstChild)
    } else {
      el.appendChild(this.element)
    }
  }

  private transitionEnd(property: string): Promise<void> {
    return new Promise((resolve) => {
      const onTransitionEnd = (e: any) => {
        if (e.propertyName === property) {
          resolve()
          this.element.removeEventListener(transitionEvent, onTransitionEnd)
        }
      }
      this.element.addEventListener(transitionEvent, onTransitionEnd)
    })
  }
}
